### DISCUSSION TITLE
Terminal settings for dosemu -t to run applications that directly access text mode video memory

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1749] [General]
`dosemu -t` allows running text mode applications with a curses-style interface. In this mode, dosemu appears to create a text mode video memory mapping that corresponds to the terminal size. That is fine for when you use a windowed graphical terminal (say, `mate-terminal`) which will often default to the well-known 80x25 size (80 columns, 25 rows).

However, when running with the interface of an application like ConnectBot, then the terminal size will often be something else that DOS applications don't expect. This results in the application's interface being torn all the way over the display. For example, if you run the following command: `insight/src$ ~/local.new/bin/dosemu -I "cpu_vm_dpmi native" -K "$PWD" -E "c:ldebug insight.com /A /E" -I "serial { com 2 device /tmp/vptty-dos }" -t` in an unusual terminal size, this is what you get from Insight (after entering a `G` command to lDebug):

![Screenshot_20220722-135054](https://user-images.githubusercontent.com/31538682/180433710-ce72cb88-98cf-4474-b1e8-94e68f5ddcb5.png)

Now, ConnectBot does have a way to force the terminal size, but if you use a terminal multiplexer like `screen` then the ConnectBot forced size will affect every tab. What I found is you can tell an individual `screen` tab to use a terminal of a certain size by running an `stty` command. In particular, that command is `stty rows 25 cols 80`.

As a bonus, you can use `stty -ixon rows 25 cols 80` to also enable Ctrl-S and Ctrl-Q to be entered to the DOS application. These two key combinations appear to be used for terminal flow control by default.

The result is an Insight that is mostly usable:

![Screenshot_20220722-135519](https://user-images.githubusercontent.com/31538682/180434222-e652782c-4f5e-438a-bba6-f4199c415ab5.png)

(Note I used the /A switch to enable ASCII-only mode, which I added to Insight. The ConnectBot terminal doesn't render all graphical drawing codepoints consistently in a monospaced way, leading to small distortions of the interface.)

**Comment by stsp** [comment]
Use `$_fixed_term_size = "80x25"`.
But you may want to fill up a bug
for you app because even in native
DOS 80x25 is not the only possible
text resolution.

**Reply by ecm-pushbx** [reply]
That works too, but there are some artefacts on the side of the display, past the screen width of the DOS. I used the switches `-t -I 'terminal { color on size "80x25" }'`. Here's what it looks like:

![Screenshot_20220722-143912](https://user-images.githubusercontent.com/31538682/180441223-925ddc9b-1748-405a-b8bd-74a96b12bb6c.png)

**Reply by ecm-pushbx** [reply]
As for the bug I believe Insight allows 80 or 90 columns (a build-time option) but can work with different amounts of rows. Eric's Terminal errors out if it detects an unsupported screen size. (With a rather nasty display bug, but that's something else.)

**Reply by andrewbird** [reply]
@ecm-pushbx, I really love the look of Insight. Do you find it useful?

**Reply by ecm-pushbx** [reply]
I've barely used it yet to be honest. My interest is mostly in trying to learn how to program a windowed interface, but I'm still very used to lDebug's command line. So it's more of an exercise in what it can do rather than something I really need or use. (Even moreso than all my other projects.)

My plan is to combine the two debuggers, and have Insight's interface control an lDebug instance across a serial port. That would allow lDebug's more sophisticated breakpoints, expression evaluator, assembler and disassembler to be used with Insight's windowed interface, either through existing or new menues and dialogs, or by dropping to a command-line window to directly control lDebug. As a bonus, the two-applications model allows for two code segments up to 64 KiB each very easily. (This goal is also why I looked into fixing Eric's Terminal, which I plan to use as the base for Insight's serial command line interface.)

Perhaps Insight or the Insight + lDebug combo will find more users than just plain lDebug. As always, I hope to create something useful or pleasing.

**Reply by stsp** [reply]
> I really love the look of Insight.

Do you plan to make dosdebug
look that way? :)

**Reply by andrewbird** [reply]
> Do you plan to make dosdebug
look that way? :)

You may laugh, but that does appeal! A while back I toyed with the idea of creating a tvision style app in qt to replace dosdebug terminal.

**Reply by stsp** [reply]
tvision in qt is probably not
the best combo as then its
no longer text-mode.
I think there are tvision ports
that are text mode.

**Reply by andrewbird** [reply]
I envisioned a qt app with a terminal pane that would hold the text based tvision interface, whilst a second pane would have an embedded sdl window holding the dosemu session, and maybe a third pane for some gdb interface. In all, far too much work.

**Reply by stsp** [reply]
> that would hold the text based tvision interface

Good so first you need to create
that text-based interface.

> In all, far too much work.

Please do the first part only - the
tvision interface. :) qt app with panes
is too much work, yeah...

**Comment by ecm-pushbx** [comment]
Using the `stty` command also enables to use [Eric's Terminal](https://hg.pushbx.org/ecm/terminal), which aborts loading otherwise. It looks like this, connected to an lDebug instance running in another dosemu2:

![Screenshot_20220722-151112](https://user-images.githubusercontent.com/31538682/180446771-5fa533f0-70ab-405f-9ef6-da700cb808ae.png)

Interestingly you can use [lClock](https://pushbx.org/ecm/web/#projects-lclock) without setting the terminal size because it only writes to what it assumes to be the end (far right) of the first line of screen memory.

![Screenshot_20220722-151329](https://user-images.githubusercontent.com/31538682/180447243-0e415170-b962-4d35-8170-01315f597d20.png)

Though as you can see, you get `15:13:290` because there's a digit zero left over from the FreeDOS kernel startup messages. It's cleaner when I do set the terminal size:

![Screenshot_20220722-151400](https://user-images.githubusercontent.com/31538682/180447722-ba4bf172-3afb-4d8b-9fac-7df6c3c63491.png)

**Reply by stsp** [reply]
> past the screen width of the DOS.

Try another terminal as there is not
supposed to be "past the screen width of the DOS"
space.

**Comment by ecm-pushbx** [comment]
Setting 50 rows modes this way also works nicely. Much less hassle than with an X window, as in https://github.com/dosemu2/dosemu2/issues/1756

**Comment by ecm-pushbx** [comment]
I encountered a different problem in using `stty` to set a different amount of columns than what ConnnectBot actually provides. This is not related to dosemu2 strictly speaking. Instead, it is bash which becomes confused about the terminal width and which makes it very difficult to edit a long scriptlet command in its line interface. When I use the right or left arrow keys to go through a long command, eventually the cursor skips parts and ends up pointing at the wrong part of the screen.

I found a method suggested [on the unix stackexchange](https://unix.stackexchange.com/questions/559386/how-can-i-revert-the-effect-of-stty-cols-n/559395#559395) which is to preserve the current amount of columns, then restore it after having run dosemu2. Thus the scriptlet I use to run Insight is now the following:

`insight/src$ COLS="$(tput cols)"; stty -ixon rows 25 cols 80; ~/local.new/bin/dosemu -I "cpu_vm_dpmi native" -K "$PWD" -E "c:ldebug.com /c='r dco4 or= f;r dco or= 4800;r int8ctrl = 0' insight.com /E /A" -I "serial { com 2 device /tmp/vptty-dos }" -t -I "serial { com 1 device /tmp/vptty-terminal-3 }"; stty cols "$COLS"; echo`

This is not entirely reliable though. But I found I can also resize the ConnectBot window, using the volume up/down buttons on the mobile. Resize it twice, once one way and then the other way. The result is the same size as we started with, but I assume it sends a `winch` signal to bash, like [I also catch in TracList](https://hg.pushbx.org/ecm/tractest/file/7cbb72094d71/traclist.pl#l451). This seems to fix bash line editing on the occasions that the `stty cols` command run after dosemu2 won't help.
